### PR TITLE
OHRM5X-374: username validation and  null validation in about API added

### PR DESCRIPTION
**Pull request by VageeshOrangeHRM**
This PR Contains following fixes:
1. OHRM5X-374: Username validation is not available.
2. About organization BE null pointer exception ( when the organization is empty or organization name is empty ) issue fixed.
3. OHRM5X-384 - Work Email unique validation not implemented.
4. New End point added for Validations (1,3).
5. Data group permission added for newly added APIs.
6. Integration test added for newly added APIs.

### CODE REVIEW

File: symfony/client/src/orangehrmAdminPlugin/pages/systemUser/SaveSystemUser.vue [hunk 1]
```diff
@@ -129,6 +130,7 @@ export default {
             (v && v.trim().length >= 5) || 'Should have at least 5 characters',
```
Comment (Super-Chama) [inline symfony/client/src/orangehrmAdminPlugin/pages/systemUser/SaveSystemUser.vue:130]: Let's refactor these with validation functions. `shouldNotExceedCharLength()`

Comment (Super-Chama) [reply symfony/client/src/orangehrmAdminPlugin/pages/systemUser/SaveSystemUser.vue:130]: _Should have at least 5 characters_
You can add a validation rule for this, if you have spare time.

File: symfony/client/src/core/util/composable/usePaginate.ts [hunk 2]
```diff
@@ -74,7 +74,7 @@ function getPageParams(pageSize: number, currentPage: number) {
 function getQueryParams(query: any) {
   const params = JSON.parse(JSON.stringify(unref(query)));
   for (const [key, value] of Object.entries(params)) {
-    params[key] = value ? value : undefined;
+    params[key] = value === null || value === '' ? undefined : value;
```
Comment (Super-Chama) [inline symfony/client/src/core/util/composable/usePaginate.ts:77]: Hope you checked this working few other places as well?

**Review by Super-Chama — COMMENTED** [review]
Change your base branch to **orangehrm/5.0-beta** and don't forget to `yarn lint`

File: symfony/plugins/orangehrmCorePlugin/Api/Rest/AboutOrganizationAPI.php [hunk 3]
```diff
@@ -60,7 +60,8 @@ public function getOne(): EndpointResult
         $aboutOrganization = new AboutOrganization();
         $employeeParamHolder = new EmployeeSearchFilterParams();
         $employeeParamHolder->setIncludeEmployees("3");
-        $organizationName = $this->getOrganizationService()->getOrganizationGeneralInformation()->getName();
+        $organization = $this->getOrganizationService()->getOrganizationGeneralInformation();
+        $organizationName = !$organization ? 'OrangeHRM' : ($organizationName = $organization->getName() === null ? 'OrangeHRM' : $organization->getName());
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmCorePlugin/Api/Rest/AboutOrganizationAPI.php:64]:
```php
$organizationName = $organization instanceof Organization ? $organization->getName() : 'OrangeHRM';
```
What do you think about this

File: symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            $isChangeableWorkEmail = $this->getEmployeeService()->getEmployeeDao()->isWorkEmailAvailableByCurrentEmail($workEmail,$employeeCurrentWorkEmail);
+        }
+        else{
+            $isChangeableWorkEmail = $this->getEmployeeService()->getEmployeeDao()->isWorkEmailAvailable($workEmail);
+        }
+        return new EndpointResourceResult(
+            ArrayModel::class,
+            [
+                self::PARAMETER_IS_CHANGEABLE_WORK_EMAIL => $isChangeableWorkEmail,
+                self::PARAMETER_WORK_EMAIL => $workEmail,
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php:63]: will remove this. no point of adding request parameter to the response. (remove const as well)

File: symfony/plugins/orangehrmAdminPlugin/Api/ValidationUserNameAPI.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    public function getOne(): EndpointResult
+    {
+        $userName = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_QUERY, self::PARAMETER_USER_NAME);
+        $isChangeableUserName = $this->getSystemUserService()->getSystemUserDao()->isUserNameExistByUserName($userName);
+        return new EndpointResourceResult(
+            ArrayModel::class,
+            [
+                self::PARAMETER_IS_CHANGEABLE_USERNAME => $isChangeableUserName,
+                self::PARAMETER_USER_NAME => $userName,
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/ValidationUserNameAPI.php:62]: will remove this. no point of adding request parameter to the response. (remove const as well)

File: symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getValidationRuleForGetOne(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(self::PARAMETER_WORK_EMAIL, new Rule(Rules::STRING_TYPE),
+                new Rule(Rules::CALLBACK, [
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php:75]: don't use Rules::CALLBACK in every places.
Is this the validation in FE as well

File: symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                function () {
+                    $workEmail = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_QUERY, self::PARAMETER_WORK_EMAIL);
+                    return filter_var($workEmail, FILTER_VALIDATE_EMAIL);
+                }
+            ])),
+            new ParamRule(CommonParams::PARAMETER_EMP_NUMBER),
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php:81]: Add validation rule, `new Rule(Rules::IN_ACCESSIBLE_EMP_NUMBERS)`

File: symfony/plugins/orangehrmAdminPlugin/Api/ValidationUserNameAPI.php [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        );
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getValidationRuleForGetOne(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(self::PARAMETER_USER_NAME, new Rule(Rules::STRING_TYPE))
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/ValidationUserNameAPI.php:73]: better to define max length

File: symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+    /**
+     * @inheritDoc
+     */
+    public function getOne(): EndpointResult
+    {
+        $empNumber = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, CommonParams::PARAMETER_EMP_NUMBER);
+        $workEmail = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_QUERY, self::PARAMETER_WORK_EMAIL);
+        $employee = $this->getEmployeeService()->getEmployeeDao()->getEmployeeByEmpNumber($empNumber);
+        $this->throwRecordNotFoundExceptionIfNotExist($employee,Employee::class);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php:51]: format this file

File: symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        $employeeCurrentWorkEmail = $employee->getWorkEmail();
+        if ($employeeCurrentWorkEmail){
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php:53]: expression not clear, better to check null

File: symfony/plugins/orangehrmPimPlugin/Dao/EmployeeDao.php [hunk 11]
```diff
@@ -445,4 +446,35 @@ public function getEmailList(): ?array {
...
+    public function isWorkEmailAvailableByCurrentEmail(string $workEmail, string $currentWorkEmail): bool
+    {
+        // this function for validating (update on validation) the work email availability. ( false -> email already exist, true - email is not exist )
+        $q = $this->createQueryBuilder(Employee::class, 'employee');
+        $q->andWhere('employee.workEmail = :workEmail');
+        $q->andWhere('employee.workEmail != :currentWorkEmail'); // we need to skip the current email on checking, otherwise count always return 1 (if current work email is not null)
+        $q->setParameter('workEmail', $workEmail);
+        $q->setParameter('currentWorkEmail', $currentWorkEmail);
+        $paginator = new Paginator($q);
+        return $paginator->count() == 0;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/Dao/EmployeeDao.php:464]:
```php
return $this->getPaginator($q)->count() === 0;
```

File: symfony/plugins/orangehrmPimPlugin/Dao/EmployeeDao.php [hunk 12]
```diff
@@ -445,4 +446,35 @@ public function getEmailList(): ?array {
...
+     * @return bool
+     */
+    public function isWorkEmailAvailable(string $workEmail): bool
+    {
+        // this function for validating the work email availability (current email is null). ( false -> email already exist, true - email is not exist )
+        $q = $this->createQueryBuilder(Employee::class, 'employee');
+        $q->andWhere('employee.workEmail = :workEmail');
+        $q->setParameter('workEmail', $workEmail);
+        $paginator = new Paginator($q);
+        return $paginator->count() == 0;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/Dao/EmployeeDao.php:478]:
```php
return $this->getPaginator($q)->count() === 0;
```
And format the file

File: symfony/plugins/orangehrmPimPlugin/config/routes.yaml [hunk 13]
```diff
@@ -488,6 +488,16 @@ apiv2_pim_update_password:
     id: 0
     _api: OrangeHRM\Pim\Api\UpdatePasswordAPI
 
+apiv2_core_validate_employee_email:
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/config/routes.yaml:491]: `apiv2_pim_validate_employee_email`

File: symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    public const PARAMETER_WORK_EMAIL = 'workEmail';
+    public const PARAMETER_IS_CHANGEABLE_WORK_EMAIL = 'valid';
+
+    /**
+     * @inheritDoc
+     */
+    public function getOne(): EndpointResult
+    {
+        $empNumber = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, CommonParams::PARAMETER_EMP_NUMBER);
+        $workEmail = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_QUERY, self::PARAMETER_WORK_EMAIL);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php:49]: Do you have any plan to handle case-sensitivity here?
e.g. abc@gmail.com, ABC@gmail.com, aBc@gmail.com 
I think all are same

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php (hunk 14 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php:49]: that is needed to be handled in the database also, Collation type utf8_general_ci to utf8_bin

File: symfony/client/src/orangehrmAdminPlugin/pages/systemUser/SaveSystemUser.vue (hunk 1 above)
Comment (VageeshOrangeHRM) [reply symfony/client/src/orangehrmAdminPlugin/pages/systemUser/SaveSystemUser.vue:130]: done

File: symfony/client/src/core/util/composable/usePaginate.ts (hunk 2 above)
Comment (VageeshOrangeHRM) [reply symfony/client/src/core/util/composable/usePaginate.ts:77]: done

File: installer/dbscript/dbscript-2.sql [hunk 15]
```diff
@@ -3619,7 +3619,8 @@ VALUES ('apiv2_admin_education', 'API-v2 Admin - Education', 1, 1, 1, 1),
        ('apiv2_admin_work_shift', 'API-v2 Admin - Work Shift', 1, 1, 1, 1),
        ('apiv2_admin_work_shift_employee', 'API-v2 Admin - Work Shift Employee', 1, 0, 0, 0),
        ('apiv2_email_subscriptions', 'API-v2 Admin - Email Subscription', 1, 0, 1, 0),
-       ('apiv2_email_subscribers', 'API-v2 Admin - Email Subscriber', 1, 1, 1, 1);
+       ('apiv2_email_subscribers', 'API-v2 Admin - Email Subscriber', 1, 1, 1, 1),
+       ('apiv2_validation_user_name', 'Admin-v2 Admin - User Name Validation', 1, 0, 0, 0);
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply installer/dbscript/dbscript-2.sql:3623]: done

File: installer/dbscript/dbscript-2.sql [hunk 16]
```diff
@@ -3739,7 +3743,8 @@ VALUES ('apiv2_pim_custom_field', 'API-v2 PIM - Custom Fields', 1, 1, 1, 1),
        ('apiv2_pim_defined_reports', 'API-v2 PIM - Defined Reports', 1, 1, 1, 1),
        ('apiv2_pim_reports', 'API-v2 PIM - Reports', 1, 0, 0, 0),
        ('apiv2_pim_reports_data', 'API-v2 PIM - Reports Data', 1, 0, 0, 0),
-       ('apiv2_pim_update_password', 'API-v2 PIM - Password Update', 0, 0, 1, 0);
+       ('apiv2_pim_update_password', 'API-v2 PIM - Password Update', 0, 0, 1, 0),
+       ('apiv2_validation_work_email', 'API-v2 PIM - Work Email Validation', 1, 0, 0, 0);
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply installer/dbscript/dbscript-2.sql:3747]: done

File: symfony/plugins/orangehrmAdminPlugin/Dao/UserDao.php [hunk 17]
```diff
@@ -373,4 +373,18 @@ public function getEmployeesByUserRole(
...
         }
     }
+
+    /**
+     * @param string $userName
+     * @return bool
+     */
+    public function isUserNameExistByUserName(string $userName): bool
+    {
+        // this function for validating the username availability. ( false -> username already exist, true - username is not exist )
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmAdminPlugin/Dao/UserDao.php:383]: done

File: symfony/plugins/orangehrmAdminPlugin/Dao/UserDao.php [hunk 18]
```diff
@@ -373,4 +373,18 @@ public function getEmployeesByUserRole(
...
+        $q = $this->createQueryBuilder(User::class, 'u');
+        $q->andWhere('u.userName = :userName');
+        $q->setParameter('userName', $userName);
+        $paginator = new Paginator($q);
+        return $paginator->count() == 0;
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmAdminPlugin/Dao/UserDao.php:388]: done

File: symfony/plugins/orangehrmAdminPlugin/Api/ValidationUserNameAPI.php (hunk 5 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmAdminPlugin/Api/ValidationUserNameAPI.php:62]: done

File: symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php (hunk 9 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php:51]: done

File: symfony/plugins/orangehrmPimPlugin/Dao/EmployeeDao.php (hunk 11 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmPimPlugin/Dao/EmployeeDao.php:464]: done

File: symfony/plugins/orangehrmPimPlugin/Dao/EmployeeDao.php (hunk 12 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmPimPlugin/Dao/EmployeeDao.php:478]: done

File: symfony/plugins/orangehrmPimPlugin/config/routes.yaml (hunk 13 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmPimPlugin/config/routes.yaml:491]: done

File: symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php (hunk 4 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php]: done

File: symfony/plugins/orangehrmAdminPlugin/Api/ValidationUserNameAPI.php (hunk 8 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmAdminPlugin/Api/ValidationUserNameAPI.php]: done

File: symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php (hunk 7 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmPimPlugin/Api/ValidationEmployeeEmailAPI.php:91]: if we added this, then, if user logged in as  ESS  (not supervisors) API will not work??

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**